base/bucket_gocb.go

```diff
@@ -1192,18 +1195,40 @@ func (bucket CouchbaseBucketGoCB) WriteUpdateWithXattr(k string, xattrKey string
 			return err
 		}
 
-		// CAS-safe write of the new value and xattr
-		LogTo("gocb", "Calling WriteCasWithXattr for key: %s cas: %d, value: %v, xattr:%v", k, cas, updatedValue, xattrMap)
-		_, err = bucket.WriteCasWithXattr(k, xattrKey, exp, cas, updatedValue, xattrMap)
+		var writeErr error
+		// If this is a tombstone, we want to delete the document and update the xattr
+		if deleteDoc {
+			// TODO: replace with a single op when https://issues.couchbase.com/browse/MB-24098 is ready
+			removeCas, removeErr := bucket.Remove(k, cas)
+			if removeErr != nil && !isRecoverableGoCBError(removeErr) {
+				LogTo("CRUD", "Remove of deleted doc during WriteUpdateWithXattr failed for key %s: %v", k, removeErr)
+				return removeErr
+			}
+			// update xattr only
+			_, writeErr := bucket.WriteCasWithXattr(k, xattrKey, exp, removeCas, nil, xattrMap)
+			if writeErr != nil && writeErr != gocb.ErrKeyExists && !isRecoverableGoCBError(writeErr) {
+				LogTo("CRUD", "Update of new value during WriteUpdateWithXattr failed for key %s: %v", k, writeErr)
+				return writeErr
+			}
+
+			// If there was no error, we're done
+			if writeErr == nil {
+				return nil
+			}
```

Comment:
But if there *is* a non-nil `writeErr`, then the code looks like it will fall down to:

```
_, writeErr = bucket.WriteCasWithXattr(k, xattrKey, exp, cas, updatedValue, xattrMap)
```

which appears to ignore the previous `writeErr` and then shadows it with the result of ` bucket.WriteCasWithXattr`